KAFKA-14960: TopicMetadata request manager

philipnee opened the pull request:
TopicMetadataRequestManager is responsible for sending topic metadata requests. The manager manages API requests and build the request accordingly. All topic metadata requests are chained, if requesting the same topic, to avoid sending requests repeatedly.

kirktrue committed:
KAFKA-14274, #3: Introduce IdempotentCloser to ensure resources are only closed once

Added IdempotentCloser to ensure resources are only closed once. While this class is broadly useful, it will be constrained to consumer internals until a later date.

kirktrue committed:
KAFKA-14274, #3: Introduce IdempotentCloser to ensure resources are only closed once

Added IdempotentCloser to ensure resources are only closed once. While this class is broadly useful, it will be constrained to consumer internals until a later date.

kirktrue committed:
KAFKA-14274, #3: Introduce IdempotentCloser to ensure resources are only closed once

Added IdempotentCloser to ensure resources are only closed once. While this class is broadly useful, it will be constrained to consumer internals until a later date.

kirktrue committed:
KAFKA-14274, #3: Introduce IdempotentCloser to ensure resources are only closed once

Added IdempotentCloser to ensure resources are only closed once. While this class is broadly useful, it will be constrained to consumer internals until a later date.

philipnee committed:
KAFKA-14960: [Part I]TopicMetadataRequestManager Implementation (#7)

TopicMetadataRequestManager implementation

---------

Co-authored-by: Lianet Magrans <lmagrans@confluent.io>

philipnee committed:
KAFKA-14960: [Part I]TopicMetadataRequestManager Implementation (#7)

TopicMetadataRequestManager implementation

---------

Co-authored-by: Lianet Magrans <lmagrans@confluent.io>

philipnee committed:
KAFKA-14960: [Part I]TopicMetadataRequestManager Implementation (#7)

TopicMetadataRequestManager implementation

---------

Co-authored-by: Lianet Magrans <lmagrans@confluent.io>

philipnee committed:
KAFKA-14960: [Part I]TopicMetadataRequestManager Implementation (#7)

TopicMetadataRequestManager implementation

---------

Co-authored-by: Lianet Magrans <lmagrans@confluent.io>

philipnee committed:
rebase

it's in a state that all tests are passing

rebase

philipnee committed:
rebase

it's in a state that all tests are passing

rebase

philipnee committed:
rebase

it's in a state that all tests are passing

rebase

philipnee committed:
rebase

it's in a state that all tests are passing

rebase

philipnee committed:
refactor existing code

philipnee committed:
refactor existing code

philipnee committed:
refactor existing code

philipnee committed:
refactor existing code

philipnee committed:
Fixing uncompleted future

philipnee committed:
Fixing uncompleted future

philipnee committed:
Revert "KAFKA-14274, #3: Introduce IdempotentCloser to ensure resources are only closed once"

This reverts commit 76d5c9641c29352ddbb896e788bfc8e189f27581.

philipnee committed:
Revert "KAFKA-14274, #3: Introduce IdempotentCloser to ensure resources are only closed once"

This reverts commit 76d5c9641c29352ddbb896e788bfc8e189f27581.

philipnee committed:
Clean up sad unused code

Update CommitRequestManager.java

clean up

clean up

remove idempotent references

philipnee committed:
Clean up sad unused code

Update CommitRequestManager.java

clean up

clean up

remove idempotent references

philipnee committed:
Update DefaultBackgroundThreadTest.java

philipnee committed:
Merge branch 'cherry-pick-topic-metadata' of github.com:philipnee/kafka into cherry-pick-topic-metadata

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -91,7 +89,7 @@ public CommitRequestManager(
 
     /**
-     * Poll for the {@link OffsetFetchRequest} and {@link OffsetCommitRequest} request if there's any. The function will
+     * Poll for the {@link OffsetFetchRequest} and {@link org.apache.kafka.common.requests.OffsetCommitRequest} request if there's any. The function will
```
Do we need to spell out the package name for OffsetCommitRequest? It doesn't seem this is done consistently.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -220,7 +214,7 @@ public NetworkClientDelegate.UnsentRequest toUnsentRequest() {
             }
 
-            OffsetCommitRequest.Builder builder = new OffsetCommitRequest.Builder(
+            org.apache.kafka.common.requests.OffsetCommitRequest.Builder builder = new org.apache.kafka.common.requests.OffsetCommitRequest.Builder(
```
Could we import the package?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/TopicMetadataApplicationEvent.java:
```diff
@@ -0,0 +1,34 @@
+
+    public String topic() {
+        return topic;
+    }
```
Do we need to add `toString`, `equals` and `hashcode`?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,210 @@
+/**
+ * <p>
+ * Manages the state of topic metadata requests. This manager returns a
+ * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.PollResult} when a request is ready to
```
Do we need to spell out the package name?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,210 @@
+ * be sent. Specifically, this manager handles the following user API calls:
+ * </p>
+ * <ul>
+ * <li>listOffsets</li>
```
Is this true? `OffsetsRequestManager` says it handles listOffsets.

philipnee replied:
🤦  - listTopics() instead.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,210 @@
+ * The manager checks the state of the {@link TopicMetadataRequestState} before sending a new one to
+ * prevent sending it without backing off from previous attempts.
+ * It also checks the state of inflight requests to avoid overwhelming the broker with duplicate requests.
+ * The {@code inflightRequests} are memoized by topic name. If all topics are requested, then {@code null} is used as the key.
```
memoized => memorized
null => Empty?

philipnee replied:
thanks - apologize for neglecting to update the doc.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,210 @@
+ * </p>
+ * <ul>
+ * <li>listOffsets</li>
+ * <li>partitionsFor</li>
```
It doesn't seem that we have wired `TopicMetadataRequestManager` to the `partitionsFor` call?

philipnee replied:
As well as listTopics() - this is addressed in a separate PR: Would you be ok if we split this into multiple PRs? If you think we should remove the documentation due to inaccurate information, I'm willing to do so.

The PR's commit is c71a18c95937dd18171f60afb4fd263ea39e9a1b - I anticipate @kirktrue will pick this into trunk soon.

junrao replied:
Yes, this is fine.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,210 @@
+                        future.complete(res);
+                        inflightRequests.remove(topic);
+                    } catch (RetriableException e) {
+                        this.onFailedAttempt(currentTimeMs);
```
This may be fine. But if a pending request is in a retry loop, but is already timed out, do we need logic to remove it from inflightRequests preemptively?

philipnee replied:
Hi Jun - Thanks for the catch - I added a test case in the TopicMetadataRequestManagerTest:146

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,210 @@
+         * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.UnsentRequest} if needed.
+         */
+        private Optional<NetworkClientDelegate.UnsentRequest> send(final long currentTimeMs) {
+            if (!this.canSendRequest(currentTimeMs)) {
```
RequestState only allows one outstanding request at a time. Does the old consumer do the same?

philipnee replied:
RequestState is a new class introduced in this consumer to handle exponential backoff. For managers that only send a single request (FindCoordinator, Heartbeat for example), the implementation is quite simple - if the current request failed, it will backoff and retry.  

For managers handling multiple requests like this one and the commit request manager - the request state exists on per request basis. I'm sure there's a better way to do this, but the original go was to offer exponential backoff functionality.

junrao replied:
It would be useful to be consistent in convention. If we want to add `this` in front of methods in the parent class, we want to do this consistently.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CommitApplicationEvent.java:
```diff
@@ -19,22 +19,23 @@
     public CommitApplicationEvent(final Map<TopicPartition, OffsetAndMetadata> offsets) {
         super(Type.COMMIT);
-        this.offsets = offsets;
-        Optional<Exception> exception = isValid(offsets);
```
`isValid` is no longer used.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CommitApplicationEvent.java:
```diff
@@ -19,22 +19,23 @@
-            throw new RuntimeException(exception.get());
+        this.offsets = Collections.unmodifiableMap(offsets);
+
+        for (OffsetAndMetadata offsetAndMetadata : offsets.values()) {
```
This is an existing issue. Could we change `toString` to include the fields in the parent class? Do we need to add `equals` and `hashcode` for consistency?

philipnee replied:
There are quite a few events out there that need to be refactored if we do it this way.  Would it be ok for you to skip this for now and I will post a MINOR patch to address after this is closed?

kirktrue replied:
If your PR goes in before mine, you can make a comment and I'll add the `toString` changes into mine.

junrao replied:
Yes, it's fine to clean this up separately.

philipnee replied:
@junrao - per your request.  Here's the patch to address your comment: https://github.com/apache/kafka/pull/14391

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -106,6 +104,7 @@ public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
             return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, Collections.emptyList());
         }
 
+        pendingRequests.inflightOffsetFetches.forEach(System.out::println);
```
Is this intended?

philipnee replied:
🤦

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -230,15 +224,20 @@ public NetworkClientDelegate.UnsentRequest toUnsentRequest() {
         public final GroupState.Generation requestedGeneration;
-        public CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> future;
-
+        CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> future;
```
Could this be private?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -381,41 +377,41 @@ public String toString() {
         List<OffsetFetchRequestState> inflightOffsetFetches = new ArrayList<>();
 
-        public boolean hasUnsentRequests() {
+        boolean hasUnsentRequests() {
```
Could this be private? Ditto below.

philipnee replied:
Sorry - I left out commit.  Fixing those obvious mistakes right away.

philipnee replied:
Sorry - leaving this protected as it is used in one of the test to verify there's no more unsent request.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -203,16 +204,17 @@ public static class UnsentRequest {
-            this.requestBuilder = requestBuilder;
-            this.node = node;
-            this.handler = handler;
+                             final BiConsumer<ClientResponse, Throwable> callback) {
```
This constructor seems unused?

philipnee replied:
I think it is according to my IDE. I think it is used in CommitRequestManager (line 219 and 256) and TopicMetadataRequestManger (line 145)

junrao reviewed (commented):
@philipnee : Thanks for the PR. Made a pass of non-testing files. Left a few comments.

philipnee committed:
Address comments on reviews.

philipnee commented:
Hey @junrao - Addressed your comments. I'm holding off on your request for `Could we change toString to include the fields in the parent class?` as I think it would be easier if I do that consistently for all other CompletableApplicationEvents. WDYT?

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,215 @@
+        CompletableFuture<Map<String, List<PartitionInfo>>> future;
+
+        public TopicMetadataRequestState(final LogContext logContext,
+                                    final Optional<String> topic,
```
indentation

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManagerTest.java:
```diff
@@ -0,0 +1,235 @@
+    private MockTime time;
+    private TopicMetadataRequestManager topicMetadataRequestManager;
+
+    private Properties props;
```
This is only used in the setup so no need to keep it at the class level

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,215 @@
+                        future.complete(res);
+                        inflightRequests.remove(topic);
+                    } catch (RetriableException e) {
+                        if (e instanceof TimeoutException) {
```
This captures the case when a single request takes long on the broker and times out. However, a separate case is that individual request returns on time but with other RetriableException. In that case, TopicMetadataRequestManager will not clean up inflightRequests even when the client request has timed out.

philipnee replied:
Hey @junrao - Sorry for writing such a long reply - I really need to brush up my memory about the network module as well as the TopicMetadataRequest...

In the [current code](https://github.com/apache/kafka/blob/b49013b73efa25466652d8d8122974e60c927ec4/clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataFetcher.java#L99-L149) I think the following is how responses should be handled.

For the case of hard failure that the callback receives a non-null exception:
We should remove the inflight request for all non-retriable and TimeoutException because we need to fail the user requests and topic the retry.  However, we should continue to retry for RetriableException until timer runs out.

For the case of soft failure (the server respond with an error code), it seems like the right thing to do is to continue to retry except for the non-retriable exception.  

It is worth noting that: If the request timed-out (API level timeout), callback is completed with TimeoutException hard failure (in the network delegate module), which should eventually causes the exception to be purged from the inflight request as mentioned above.

To your question: if the responses include a retriable exception, I believe the right thing to do is to continue to retry until the timer runs out.

In fact - it seems like `if (e instanceof TimeoutException)` (that I added in the last commit) is actually incorrect.  Because it is a RetriableException at it should be retried per:

```
if (error.exception() instanceof RetriableException)
                            shouldRetry = true;
```

What's missing here is handling the hard RetriableException and TimeoutException.  I will fix the code and add test cases to cover the scenarios above.

junrao commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadTest.java:
```diff
@@ -287,14 +298,13 @@ public void testAssignmentChangeEvent() {
         verify(commitManager, times(1)).updateAutoCommitTimer(currentTimeMs);
         verify(commitManager, times(1)).maybeAutoCommit(offset);
-
-        backgroundThread.close();
```
It seems that we close backgroundThread in other cases?

junrao commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManagerTest.java:
```diff
@@ -0,0 +1,235 @@
+        try {
+            future.get();
+        } catch (Throwable e) {
+            System.out.println(e.getMessage());
```
In general, we should use logging instead of print.

junrao reviewed (commented):
@philipnee : Thanks for the updated PR. Made a pass of all files. A few more comments.

philipnee committed:
PR comments

Address PR

Update TopicMetadataRequestManager.java

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -141,35 +142,57 @@ private Optional<NetworkClientDelegate.UnsentRequest> send(final long currentTim
+                handleException(exception, responseTimeMs);
+                return;
+            }
+            handleResponse(response, responseTimeMs);
```
It seems that the exception handling logic here ends up in 2 places. Apart from the initial `handleException`, this `handleResponse` could be as well end up handling exceptions, in a very similar way (catch block on `handleResponse` vs the `handleException`). If we prefer to keep them separated because of the "`retryOnTimeout`" part, which seems to be what makes them different, it would be helpful maybe to add a comment to describe that

philipnee replied:
thanks, @lianetm - Do you think it would be clearer to keep them in 1 place? The intention was to handle hard failures in handleException and soft failures (i.e. returning an error code) in handleResponse.

lianetm replied:
I think it would be clearer and better to maintain/troubleshoot if error handling is in one place. Maybe having a single `handleException(final Throwable exception, final long responseTimeMs, boolean retryOnTimeout)`, called in both cases:
```
if (exception != null) {
                // Handle hard errors
                handleException(exception, responseTimeMs, false);
                ...
}
...
catch (Exception e) {
                // Handle soft errors
                handleException(e, responseTimeMs, true);
}
```

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,238 @@
+
+        private void processResponseOrException(final ClientResponse response,
+                                                final Throwable exception) {
+            long responseTimeMs = System.currentTimeMillis();
```
Could we pass in currentTime from the caller?

philipnee replied:
Hi Jun - understood making repeated system calls is bad.  However - we do need to get the time when the response is received, so this is kind of the lazy way to do it.  

I think it is possible to expose the time during the callback/future completion during the network poll (because it is computed at the top of the background thread loop and pass down to network poll.

junrao replied:
Actually, could we use `ClientResponse.receivedTimeMs`?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,238 @@
+        }
+
+        private void handleException(final Throwable exception, final long responseTimeMs) {
+            if (exception instanceof TimeoutException || !(exception instanceof RetriableException)) {
```
TimeoutException is a RetriableException. Should we just treat it as other RetriableException? That seems to be what the old consumer does.

lianetm replied:
Just careful here because the current `TopicMetadataFetcher` handles 2 timeout cases differently regarding retries:
- TimeoutException contained in a MetadataResponse as a response error is retried [here](https://github.com/apache/kafka/blob/f46db86d34f9e5fe1b0d7604306a5108a89c113e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataFetcher.java#L133)
- TimeoutException in getting the actual MetadataResponse throws exception [here](https://github.com/apache/kafka/blob/f46db86d34f9e5fe1b0d7604306a5108a89c113e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataFetcher.java#L151)

philipnee replied:
I think for now - the manager will continue to retry all RetriableException including TimeoutException. Currently, we don't have a way to cancel the request when the user API request times-out, so this is a gap there, but a jira is filed here: https://issues.apache.org/jira/browse/KAFKA-15475

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,238 @@
+    private final LogContext logContext;
+
+    public TopicMetadataRequestManager(final LogContext logContext, final ConsumerConfig config) {
+        this.logContext = logContext;
```
Could we just get rid of `this`? Ditto below.

philipnee replied:
sorry - got into the habit of using `this`. 😅

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,238 @@
+                                         final long retryBackoffMaxMs) {
+            super(logContext, TopicMetadataRequestState.class.getSimpleName(), retryBackoffMs,
+                retryBackoffMaxMs);
+            this.future = new CompletableFuture<>();
```
Could we just get rid of `this`?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,238 @@
+                future.complete(res);
+                inflightRequests.remove(topic);
+            } catch (RetriableException e) {
+                onFailedAttempt(responseTimeMs);
```
The case that I am wondering is the following.

DefaultBackgroundThread is in a retry loop for an inflightRequest. PrototypeAsyncConsumer.partitionsFor eventually calls CompletableApplicationEvent.get(Timer timer) and times out. Do we still keep the request in inflightRequests at that point? Is that a concern?

philipnee replied:
Thanks Jun for raising this point - I think what we should do is to complete the request with a non-retriable exception to prevent the request from retrying. Otherwise - the request manager will never remove the request from the inflightRequest.

junrao reviewed (commented):
@philipnee : Thanks for the updated PR. A few more comments.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -298,8 +289,8 @@ private void onFailure(final long currentTimeMs,
         private void retry(final long currentTimeMs) {
-            onFailedAttempt(currentTimeMs);
-            onSendAttempt(currentTimeMs);
+            this.onFailedAttempt(currentTimeMs);
```
currentTimeMs is when the request is sent, not when the response is receive. Could we pass along `ClientResponse.receivedTimeMs`?

philipnee replied:
Thanks for catching this.

philipnee replied:
Addressed, see line #259

philipnee committed:
PR comment - passing responseTime from ClientResponse

philipnee committed:
addressed more comments

philipnee committed:
Merge branch 'trunk' into cherry-pick-topic-metadata

philipnee committed:
clean up

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorRequestManager.java:
```diff
@@ -38,8 +38,8 @@
  * Whether there is an existing coordinator.
  * Whether there is an inflight request.
  * Whether the backoff timer has expired.
- * The {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.PollResult} contains either a wait timer
```
just cleaning up long references

philipnee committed:
clean up

philipnee commented:
@junrao @lianetm - I hope I've addressed most, if not all, comments that you pointed out. So, thank you both for reviewing my code.

We did have an offline discussion about user API timeout handling, and a bug report is filed here [KAFKA-15475](https://issues.apache.org/jira/browse/KAFKA-15475).  I will submit a patch after this PR to address this small gap.

Lastly, here are the failing tests:

```
Build / JDK 20 and Scala 2.13 / testReplication() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest
53s
Build / JDK 17 and Scala 2.13 / testMultiWorkerRestartOnlyConnector – org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest
2m 24s
Build / JDK 17 and Scala 2.13 / testCreateTokenForOtherUserFails(String).quorum=kraft – kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest
9s
Build / JDK 11 and Scala 2.13 / testConfigurationOperations() – org.apache.kafka.controller.QuorumControllerTest
<1s
Build / JDK 11 and Scala 2.13 / executeTieredStorageTest(String).quorum=zk – org.apache.kafka.tiered.storage.integration.EnableRemoteLogOnTopicTest
43s
Build / JDK 8 and Scala 2.12 / testMultiWorkerRestartOnlyConnector – org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest
2m 25s
```

I believe these are flaky.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,233 @@
+
+    public TopicMetadataRequestManager(final LogContext context, final ConsumerConfig config) {
+        logContext = context;
+        log = logContext.logger(this.getClass());
```
Could we remove `this`?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -0,0 +1,233 @@
+        }
+
+        private void completeFutureAndRemoveRequest(final Throwable throwable) {
+            this.future.completeExceptionally(throwable);
```
Could we remove `this`?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -91,7 +90,7 @@ public CommitRequestManager(
 
     /**
-     * Poll for the {@link OffsetFetchRequest} and {@link OffsetCommitRequest} request if there's any. The function will
+     * Poll for the {@link OffsetFetchRequest} and {@link OffsetCommitRequestState} request if there's any. The function will
```
Hmm, OffsetCommitRequestState should be OffsetCommitRequest, right?

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -395,27 +386,28 @@ class PendingRequests {
         List<OffsetFetchRequestState> inflightOffsetFetches = new ArrayList<>();
 
-        public boolean hasUnsentRequests() {
+        // Visible for teseting
```
typo teseting

junrao reviewed (commented):
@philipnee : Thanks for the updated PR. Just a few minor comments.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -395,27 +386,28 @@ class PendingRequests {
-         *  <p>If the request is new, it invokes a callback to remove itself from the {@code inflightOffsetFetches}
-         *  upon completion.</>
+         * <p>If the request is new, it invokes a callback to remove itself from the {@code inflightOffsetFetches}
+         * upon completion.</>
```
nit: wrong tag </> and other unclosed ones above

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -451,11 +443,11 @@ private CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> addOffsetFetch
          * Clear {@code unsentOffsetCommits} and moves all the sendable request in {@code unsentOffsetFetches} to the
          * {@code inflightOffsetFetches} to bookkeep all of the inflight requests.
-         *
+         * <p>
```
was the intention here to add an extra line in-between? should be `<p/>` if so

philipnee replied:
😓 - must be the residual from some edits..

philipnee committed:
Update CommitRequestManager.java

Updates based on review comments

philipnee commented:
@junrao @lianetm - Thanks for another round of reviews. The latest build passed with the following test failures

```
Build / JDK 17 and Scala 2.13 / testBecomeFollowerWhileOldClientFetchInPurgatory() – kafka.server.ReplicaManagerTest
<1s
Build / JDK 17 and Scala 2.13 / [1] enableRemoteStorage=true – kafka.server.ReplicaManagerTest
<1s
Build / JDK 20 and Scala 2.13 / testRackAwareRangeAssignor() – integration.kafka.server.FetchFromFollowerIntegrationTest
46s
Build / JDK 20 and Scala 2.13 / testBumpTransactionalEpoch(String).quorum=zk – kafka.api.TransactionsTest
13s
Build / JDK 20 and Scala 2.13 / testCommitTransactionTimeout(String).quorum=kraft – kafka.api.TransactionsTest
9s
Build / JDK 20 and Scala 2.13 / executeTieredStorageTest(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.DeleteTopicTest
41s
Build / JDK 11 and Scala 2.13 / testCloseDuringRebalance() – kafka.api.ConsumerBounceTest
10s
Build / JDK 11 and Scala 2.13 / [2] Type=Raft-Isolated, Name=testDescribeQuorumReplicationSuccessful, MetadataVersion=3.7-IV0, Security=PLAINTEXT – org.apache.kafka.tools.MetadataQuorumCommandTest
1m 8s
Build / JDK 8 and Scala 2.12 / testMultiWorkerRestartOnlyConnector – org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest
2m 31s
Build / JDK 8 and Scala 2.12 / testBalancePartitionLeaders() – org.apache.kafka.controller.QuorumControllerTest
```
I believe they are irrelevant.

junrao reviewed (approved):
@philipnee : Thanks for the latest PR. LGTM
